### PR TITLE
Cache improvements

### PR DESCRIPTION
**Pull request by kumavis**
adds tests around `cache`, `cache-utils`, and `inflight-cache`

solves problem with inflight cache misgenerating cacheId

### CODE REVIEW

File: test/cache.js [hunk 1]
```diff
@@ -86,13 +86,59 @@ cacheTest('getCode for an unspecified block, then for the latest, should return
...
+
+cacheTest('getBlockForNumber for 0 then block 1', [{
+  method: 'eth_getBlockByNumber',
+  params: ['0x0'],
+}, {
+  method: 'eth_getBlockByNumber',
+  params: ['0x1'],
+}], false)
+
+cacheTest('getBlockForNumber for block 2', [{
```
Comment (flyswatter) [inline test/cache.js]: This test is mis-labeled

File: test/index.js [hunk 2]
```diff
@@ -1,7 +1,12 @@
...
+require('./cache-utils')
 require('./cache')
+require('./inflight-cache')
 require('./filters')
 require('./solc')
 require('./wallet')
 require('./subproviders/sanitizer')
 require('./subproviders/vm')
+// require('./subproviders/ipc')
+// require('./subproviders/etherscan')
```
Comment (flyswatter) [inline test/index.js]: Why deactivate these previously passing tests?

**Comment by flyswatter** [comment]
Don't ask me why bumping node got that test passing.

File: test/cache.js (hunk 1 above)
Comment (flyswatter) [inline test/cache.js]: When I correct the test to match the label, it fails locally.

```
# cache - getBlockForNumber for block 2
ok 190 did not error
ok 191 has response
ok 192 cacheProvider did see "eth_getBlockByNumber"
ok 193 cacheProvider did NOT handle "eth_getBlockByNumber"
ok 194 handlingProvider did see "eth_getBlockByNumber"
ok 195 handlingProvider did handle "eth_getBlockByNumber"
ok 196 did not error
ok 197 has response
ok 198 cacheProvider did see "eth_getBlockByNumber"
not ok 199 cacheProvider did handle "eth_getBlockByNumber"
  ---
    operator: equal
    expected: 1
    actual:   0
    at: Web3ProviderEngine._inspectResponseForNewBlock (/Users/danfinlay/Documents/Development/ethereum/metamask/node_modules/web3-provider-engine/index.js:164:12)
  ...
not ok 200 handlingProvider did NOT see "eth_getBlockByNumber"
  ---
    operator: equal
    expected: 1
    actual:   2
    at: Web3ProviderEngine._inspectResponseForNewBlock (/Users/danfinlay/Documents/Development/ethereum/metamask/node_modules/web3-provider-engine/index.js:164:12)
  ...
not ok 201 handlingProvider did NOT handle "eth_getBlockByNumber"
  ---
    operator: equal
    expected: 1
    actual:   2
    at: Web3ProviderEngine._inspectResponseForNewBlock (/Users/danfinlay/Documents/Development/ethereum/metamask/node_modules/web3-provider-engine/index.js:164:12)
  ...
```

File: test/index.js (hunk 2 above)
Comment (kumavis) [inline test/index.js]: they were not recently being run as part of the test suite, and they didnt pass on one machine
( one required a running eth node, the other hits etherscan and didnt complete )

**Pull request merged**